designs/0019-variadic_odes.md

```diff
@@ -0,0 +1,173 @@
+- Feature Name: Variadic arguments for ODE solvers
+- Start Date: 2020-04-20
+- RFC PR: 19
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+This proposal is to make a set of ODE solvers that take a variadic list of
+arguments.
+
+# Motivation
+[motivation]: #motivation
+
+The goal is to avoid the argument packing and unpacking that the current solvers
+depend on.
+
+Argument packing is inconvenient, makes code hard to read, and introduces a lot
+of possible errors.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For the bdf solver, I am proposing introducing two new functions: `ode_bdf` and
+`ode_tol_bdf` (there will be two functions similarly introduced for the rk45 and
+adams solvers).
+
+The first is a BDF solver with default tolerance settings, and the second
+requires the tolerances to be set manually. This is different from the current
+solvers where tolerances are presented as overloads to the same function name.
+With the variadic argument this isn't possible anymore so the function with
+tolerances is broken out separately. The difference between `ode_bdf` and
+`ode_bdf_tol` is described later. For now, just focus on `ode_bdf` (where
+there are no tolerance arguments).
+
+The proposed `ode_bdf` interface is:
+```
+vector[] ode_bdf(F f,
+                 vector y0,
+                 real t0, vector times,
+                 T1 arg1, T2 arg2, ...)
+```
+
+The arguments are:
+1. ```f``` - User-defined right hand side of the ODE (`dy/dt = f`)
+2. ```y0``` - Initial state of the ode solve (`y0 = y(t0)`)
+3. ```t0``` - Initial time of the ode solve
+4. ```times``` - Sequences of times to which the ode will be solved (each
+  element must be greater than t0).
+5. ```arg1, arg2, ...``` - Arguments passed unmodified to the ODE right hand
+  side (`f`). Can be any non-function type.
+
+In this proposal, the user-defined ODE right hand side interface is:
+
+```
+vector f(real t, vector y, T1 arg1, T2 arg2, ...)
+```
+
+The arguments are:
+1. ```t``` - Time at which to evaluate the ODE right hand side
+2. ```y``` - State a which to evaluate the ODE right hand side
+3. ```arg1, arg2, ...``` - Arguments pass unmodified from the `ode_bdf` call.
+  The types `T1`, `T2`, etc. need to match between this function and the
+  `ode_bdf` call.
+
+A call to `ode_bdf` returns the solution of the ODE specified by the right hand
+side (`f`) and the initial conditions (`y0` and `t0`) at the time points given
+by the `times` argument. The solution is given by an array of vectors.
+
+The proposed `ode_bdf_tol` interface is:
+```
+vector[] ode_bdf_tol(F f,
+                     vector y0,
+                     real t0, vector times,
+                     real rel_tol, real abs_tol,
```

Comment:
> From what I have seen it can be beneficial to have more accuracy for the state, which you can relax the accuracy for the gradients.

It would be nice to have optional arguments. I kinda lean yes to exposing things. Without doing that we won't be able to experiment either. This will probably also change with the adjoint stuff. I think something for later.

Reply:
So, can you expose that functionality from CVODES so that we can test it a bit?

Reply:
Yeah sure, we can try it